theories/crypto/KeyEncapsulationMechanismsROM.eca

```diff
@@ -0,0 +1,1476 @@
+  *before* the stage in which it is expected to distinguish/return a break.
+  Hence, the adversary is able to produce encapsulations
+  *and* query for decryptions of chosen ciphertexts.
+*)
+(** Interface for oracles employed in CCA1 security games in ROM **)
+module type Oracles_CCA1i_ROM (RO : RandomOracle) (S : Scheme) = {
+  proc init(sk_init : sk_t) : unit
+  proc decaps(c : ctxt_t) : key_t option
+}.
+
+
+(* 
+  adaptive Chosen-Ciphertext Attacks (Traditional: CCA2, Modern : CCA) in ROM.
+  The adversary is given the considered public key and access to a decryption oracle throughout.  
+  Hence, the adversary is able to produce encapsulations
+  *and* query for decryptions of chosen ciphertexts (potentially barring ciphertexts
+  that are part of the challenge).
+  Traditionally, this was analogous to CCA2 security for PKE schemes, meaning there were
+  two adversary stages: one before receiving the challenge (given a public key and access to a 
+  non-restricted decapsulation oracle), and one after receiving the challenge (given access to a
+  restricted decapsulation oracle, i.e., one that prohibited querying the challenge).
+  Over time, the formalization shifted toward only considering the second adversary stage 
+  (provding the public key(s) to this stage as well).
+  Here, we denote the traditional one by CCA2 (as we do for PKE schemes), and the modern one by CCA.
+*)
+(** Interface for oracles employed in CCA2 (CCA) security games in ROM **)
+module type Oracles_CCA2i_ROM (RO : RandomOracle) (S : Scheme) = {
+  proc init(sk_init : sk_t, c'_init : ctxt_t) : unit
+  proc decaps(c : ctxt_t) : key_t option
+}.
+
+
+(* 
+  One-Wayness (OW) in ROM.
+  The adversary is asked to produce the (symmetric) key
+  encapsulated by a given ciphertext.
+*)
+(*
+  One-Wayness under Chosen-Plaintext Attacks (OW-CPA) in ROM.
+  In a CPA setting, the adversary is asked to produce the (symmetric) key
+  encapsulated by a given ciphertext.
+*)
+(** Adversary class considered for OW-CPA in ROM **)
+module type Adv_OWCPA_ROM (RO : RandomOracle) = {
+  include Adv_OWCPA
+}.
+
+(** OW-CPA security game in ROM **)
+module OW_CPA_ROM (RO : RandomOraclei) (S : Scheme_ROM, A : Adv_OWCPA_ROM) = {
+  proc main() : bool = {
+    var r : bool;
+    
+    RO.init();
+    
+    r <@ OW_CPA(S(RO), A(RO)).main();
+    
+    return r;
+  }
+}.
+
+
+(*
+  One-Wayness under non-adaptive Chosen-Ciphertext Attacks (OW-CCA1) in ROM.
+  In a CCA1 setting, the adversary is asked to produce the (symmetric) key
+  encapsulated by a given ciphertext.
+*)
+(** Adversary class considered for OW-CCA1 in ROM **)
+module type Adv_OWCCA1_ROM (RO : RandomOracle) (O : Oracles_CCA) = { 
+  proc scout(pk : pk_t) : unit { O.decaps } 
+  proc find(c : ctxt_t) : key_t { }
+}.
+
+(** OW-CCA1 security game in ROM **)
+module OW_CCA1_ROM (RO : RandomOraclei) (S : Scheme_ROM, O : Oracles_CCA1i_ROM, A : Adv_OWCCA1_ROM) = {
```

Comment:
Consistency of functor parameter declaration?

Reply:
Right, this is actually something that I wasn't sure about (and have probably been inconsistent with). Currently, in terms of formatting functor parameters, I have `(RandomOracle) (Everything, From, Standard, Model)` in the ROM theory. That is, unless things do not fit on a single line, because in that case I group everything by type (see, e.g., `ANO_CCA2_ROM`). In the non-ROM theory, I group everything together if it fits on a single line, and again group by type if it doesn't fit on a single line. Perhaps simply always grouping things by type is the most consistent?

Reply:
If there's a principle behind it, it's fine. The fact that we have two syntactic ways of expressing the same thing is at fault here, not your usage of it. (We need the space-separated one, so if one goes away to simplify the parser, it will be comma-separation.)